[IPPInAppFeedback] Banner dismiss logic

iamgabrielma opened the pull request:
Closes: #8589
Part of https://github.com/woocommerce/woocommerce-ios/issues/8530

## Description

This PR adds the logic to dismiss the IPP feedback banner.

## TODO

- [x] Dismisses the IPP banner upon tapping
- [x] Presents confirmation screen upon dismissal
- [x] Dismisses the IPP banner upon survey completion.

## Changes
- Updates the `setFeatureAnnouncementDismissed` AppSettingsStore/AppSettingsAction to accept a new `remindAfter` parameter, this allows us to specify a number of days for the reminder to be triggered (previous default to 14)
- Adds an `UIAlertController` to be shown when the IPP banner is dismissed, with 2 buttons: "Remind me later", and "Don't show again".
- Once the IPP banner is dismissed, we update the feedback status to `.dismissed`, and the feedback visibility to be reminded later or never.
- Updated, and added Unit Tests


## Testing instructions
- Go to My Store > Orders and see the IPP feedback banner appear
- Tap on "x" and see a confirmation alert with 2 options: "remind me later", and "Don't show again"
- Confirm that the banner dismisses no matter the option you tap
- Note: If Order Creation feedback banner should appear in your store, this will show after we've dismissed the IPP banner.

![Simulator Screen Recording - iPhone 11 Pro - 2023-01-19 at 10 19 40](https://user-images.githubusercontent.com/3812076/213340206-f20a6397-e0c5-473a-8eeb-dde96bd53b7c.gif)

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->


---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

iamgabrielma committed:
Add IPP banner visibility state as a new publisher

iamgabrielma committed:
Add initial dismiss alert using ActionSheet

iamgabrielma committed:
Make UIAlertController style to display as alert

iamgabrielma committed:
Update banner status as dismissed, remind later, or never

iamgabrielma committed:
Merge branch 'trunk' into issue/8589-ipp-banner-dismiss

# Conflicts:
#	WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift
#	WooCommerce/Classes/ViewRelated/Orders/OrderListViewModel.swift

iamgabrielma committed:
Remove trailing whitespace

iamgabrielma committed:
Add remindAfter param to setFeatureAnnouncementDismissed action

iamgabrielma committed:
Unit Test new remindAfter param

iamgabrielma committed:
Fix Line Length Violation

iamgabrielma commented:
👋  @toupper I'd appreciate an "early" review on this one before is fully ready for review, as I'm unsure how to deal with the case of "Dismisses the IPP banner upon survey completion", happy to get any pointer on what would work for this case. For context:

- If I use the existing `SurveyCoordinatingController` (current solution), we get the WooCommerce navigation flow, but we can't access to any completion handler to work with when the survey is submitted.
- If I use the existing `SurveyViewController` we get a completion handler that is effectively called when the survey is submitted, but we lose the WooCommerce flow and get the Crowdsignal one instead.

We seem to use the first option across other surveys, but I'm failing to understand how are we dealing with the logic that needs to happen when a user submits the survey from the modal view. This will be needed for tracking FeedbackStatus as well.

Edit: This is resolved with 523fe0a8ef8ecf31a8b08612138a2b7d5acc73d4  (Reference: p1674118095143339?thread_ts=1674101250.398509&cid=CGPNUU63E-slack-CGPNUU63E). We'll be dismissing the banner right away upon tapping, and then we'll figure out if we need to show it again or not via checking the FeedbackStatus

iamgabrielma committed:
Add dismiss upon tapping on survey button

toupper commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift:
```diff
@@ -796,14 +796,19 @@ private extension OrderListViewController {
     private func createIPPFeedbackTopBanner() -> TopBannerView {
         let shareIPPFeedbackAction = TopBannerViewModel.ActionButton(title: Localization.shareFeedbackButton, action: { _ in
             self.displayIPPFeedbackBannerSurvey()
```
Do we need to add `[weak self]` for these references? Also for other `self` usage within a block in this PR

iamgabrielma replied:
We definitely do 🙇 , added both on 70d9cbdf2680cba8f3b1f3624b54d678c19748e6

toupper commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift:
```diff
@@ -816,6 +821,34 @@ private extension OrderListViewController {
         self.present(surveyNavigation, animated: true, completion: nil)
     }
+
+    private func dismissIPPFeedbackBannerSurvey() {
```
In order to be more precise, I would rename this function to actually reflect what's happening inside. Perhaps something like `func showIPPFeedbackDismissAlert()` ?

iamgabrielma replied:
Yes, the method internals changed but I didn't change the name. Modified on 909bea5b2ca4132b493b880c77a5b9fde9f8adfe

toupper commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift:
```diff
@@ -867,6 +911,7 @@ private extension OrderListViewController {
         static let estimatedHeaderHeight = CGFloat(43)
         static let estimatedRowHeight = CGFloat(86)
         static let placeholderRowsPerSection = [3]
+        static let remindAfterDays = 7
```
I would be more specific here, what is it that we are using this value for? IPP Feedback or something else? Also, this pertains to view model logic, as mentioned before.

toupper commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift:
```diff
@@ -816,6 +821,34 @@ private extension OrderListViewController {
+    }
+
+    private func remindMeLaterTapped() {
+        viewModel.dismissIPPFeedbackBanner(remindLater: true, remindAfter: Settings.remindAfterDays)
```
The `remindLater` and `remindAfter` logic seem to fit better in the view model logic, I would move that there while keeping the UI actions here.

iamgabrielma replied:
That makes sense, moved to the view model on https://github.com/woocommerce/woocommerce-ios/pull/8657/commits/80a1b1b5043c031848afcf0ae379550cf69bf84a

toupper commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewModel.swift:
```diff
@@ -196,6 +196,23 @@ final class OrderListViewModel {
         stores.dispatch(action)
     }
 
+    func dismissIPPFeedbackBanner(remindLater: Bool, remindAfter: Int?) {
```
As mentioned before, `remindLater` and `remindAfter` shouldn't be exposed to the view, but kept here inside the view model logic. We can also test it that way.

toupper commented on Yosemite/Yosemite/Stores/AppSettingsStore.swift:
```diff
@@ -178,8 +178,8 @@ public class AppSettingsStore: Store {
             loadCouponManagementFeatureSwitchState(onCompletion: onCompletion)
-        case .setFeatureAnnouncementDismissed(campaign: let campaign, remindLater: let remindLater, onCompletion: let completion):
-            setFeatureAnnouncementDismissed(campaign: campaign, remindLater: remindLater, onCompletion: completion)
+        case .setFeatureAnnouncementDismissed(campaign: let campaign, remindLater: let remindLater, remindAfter: let remindAfter, onCompletion: let completion):
```
I would encapsulate `remindLater` and `remindLater` in one value, I don't think we need both here. If we have both we can give place to inconsistent states, e.g., `remindLater` false and `remindAfter` to 5 days. For instance, we can solve this as follows:

- We can have only `remindAfter`, where 0 means it shouldn't be reminded. Or,
- `remindAfter` can be an optional, being `nil` meaning it shouldn't be reminded. I prefer this one.

iamgabrielma replied:
Much cleaner indeed, changes here: https://github.com/woocommerce/woocommerce-ios/pull/8657/commits/217758491434831705a861af3c0582da903af916 Now we have a unique Optional parameter `remindAfterDays` rather than two.

toupper commented on Yosemite/Yosemite/Stores/AppSettingsStore.swift:
```diff
@@ -178,8 +178,8 @@ public class AppSettingsStore: Store {
             loadCouponManagementFeatureSwitchState(onCompletion: onCompletion)
-        case .setFeatureAnnouncementDismissed(campaign: let campaign, remindLater: let remindLater, onCompletion: let completion):
-            setFeatureAnnouncementDismissed(campaign: campaign, remindLater: remindLater, onCompletion: completion)
+        case .setFeatureAnnouncementDismissed(campaign: let campaign, remindLater: let remindLater, remindAfter: let remindAfter, onCompletion: let completion):
```
If keep `remindAfter`, it would be valuable to specify what time unit are we expecting: `remindAfterDays`

toupper commented:
Great job @iamgabrielma! I left some comments there, let me know what you think of that. Once we are done with that, I will test the flow 👍

toupper commented on Yosemite/YosemiteTests/Stores/AppSettingsStoreTests.swift:
```diff
@@ -877,6 +877,24 @@ extension AppSettingsStoreTests {
         XCTAssert(Calendar.current.isDate(actualRemindAfter, inSameDayAs: twoWeeksTime))
     }
 
+    func test_setFeatureAnnouncementDismissed_with_remindAfter_seven_days_stores_reminder_date_in_one_week() throws {
```
It would be great to use the agreed test function naming syntax here as well:

`test_setFeatureAnnouncementDismissed_when_remindAfter_is_seven_days_stores_reminder_then_date_saved_date_is_one_week`

iamgabrielma replied:
Yes, you're right! Changed f8a2586b813a1a49f3ff0d4f6943c3aa82661b28

iamgabrielma committed:
Add missing weak references to self in closures

iamgabrielma committed:
Rename method

iamgabrielma committed:
Rename test method for consistency

iamgabrielma committed:
Move dismiss logic to view model

iamgabrielma committed:
Modify setFeatureAnnouncementCampaign action

We’re modifying the action by removing the parameters remindLater, and remindAfter in favor of an unique remindAfterDays parameter. If this is nil, no reminder is set. Otherwise the number of days will be set.

iamgabrielma committed:
Simplify dismissIPPFeedbackBanner to use only remindAfterDays

iamgabrielma committed:
Add comment to NSLocalizedString

iamgabrielma committed:
Resolve storeDismissedSetting with Optional remindAfterDays

By making remindAfterDays an Optional Int, we need to edit storeDismissedSetting as well to either save the current date if is true (by passing 0, so adds 0 days to the current date) or nil if is false

iamgabrielma commented on Yosemite/YosemiteTests/Stores/AppSettingsStoreTests.swift:
```diff
@@ -841,14 +841,31 @@ final class AppSettingsStoreTests: XCTestCase {
+    func test_setFeatureAnnouncementDismissed_for_campaign_when_remindAfterDays_is_nil_then_does_not_store_current_date() throws {
+        // Given
+        try fileStorage?.deleteFile(at: expectedGeneralStoreSettingsFileURL)
+        // When
```
Nit: Add extra space.

iamgabrielma commented on Yosemite/YosemiteTests/Stores/AppSettingsStoreTests.swift:
```diff
@@ -859,14 +876,15 @@ extension AppSettingsStoreTests {
     }
 
-    func test_setFeatureAnnouncementDismissed_with_remindLater_true_stores_reminder_date_in_two_weeks() throws {
+    func test_setFeatureAnnouncementDismissed_with_remindAfterDays_two_weeks_stores_reminder_date_in_two_weeks() throws {
```
```suggestion
    func test_setFeatureAnnouncementDismissed_when_remindAfterDays_two_weeks_then_stores_reminder_date_in_two_weeks() throws {
```

toupper replied:
`test_setFeatureAnnouncementDismissed_when_remindAfterDays_is_two_weeks_then_stores_reminder_date_is_two_weeks()`

iamgabrielma commented:
Thanks for the review @toupper ! I made the changes and this is ready for review 🙇

toupper commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewModel.swift:
```diff
@@ -351,6 +351,34 @@ final class OrderListViewModel {
+// MARK: - In-Person Payments Feedback Banner
+
+extension OrderListViewModel {
+    func dismissIPPFeedbackBanner(remindAfterDays: Int?) {
```
Nice change moving the logic to the view model! However, the `remindAfterDays` logic can be contained in the view model; the view doesn't need to know about when are we going to show it again. With that, we can remove this parameter `remindAfterDays` from this function.

iamgabrielma replied:
Changed here the remaining use of `remindAfterDays`: https://github.com/woocommerce/woocommerce-ios/pull/8657/commits/d5e02a27e6a954cb33fb0ed65709996599de07b5 to a new method, so this is entirely contained in the view model now 👍

toupper commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewModel.swift:
```diff
@@ -351,6 +351,34 @@ final class OrderListViewModel {
+        stores.dispatch(updateBannerVisibility)
+    }
+
+    func remindMeLaterTapped() {
```
We can name this function more specifically for the IPP Feedback Banner.

iamgabrielma replied:
Renamed on https://github.com/woocommerce/woocommerce-ios/pull/8657/commits/9884f22a964bcc50e635af902154bd427e2798da

iamgabrielma committed:
Contain dismissal logic to view model only

iamgabrielma committed:
Rename test for consistency

iamgabrielma committed:
Rename dismiss methods to be more explicit

toupper commented:
Great job! LGTM :shipit:

iamgabrielma committed:
Merge branch 'trunk' into issue/8589-ipp-banner-dismiss
